File: SearchSortedArrayUnkownSize.java
//Time Complexity: O(logn+logm)
//Space Complexity: O(1)

class SearchSortedUnknownSize {
    public int search(ArrayReader reader, int target) {
        int low = 0, high = 10;

        while(target > reader.get(high)){
            low = high+1;
            high = high*2;
        }

        while(low <= high){
            int mid = low+(high-low)/2;
            int middle = reader.get(mid);

            if(middle == target){
                return mid;
            }else if(target < middle){
                high = mid-1;
            }else{
                low = mid+1;
            }
        }
        return -1;
    }
}

File: searchsortedarrayunkownsize.py
# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
#class ArrayReader:
#    def get(self, index: int) -> int:

class Solution:
    def search(self, reader: 'ArrayReader', target: int) -> int:
        
        if reader.get(0)==target:
            return 0
        
        left,right=0,1
        while reader.get(right)<target:
            left=right
            right=right*2
            #print(right)
        
        while left<=right:
            #mid=left+right//2 ---->Gives time limit exceed
            mid=left+(right-left)//2
            val=reader.get(mid)
            
            if val==target:
                return mid
            if val>target:
                right=mid-1
            else:
                left=mid+1
        return -1